### Car Price Prediction Assignment 2

In [1]:
# import required libraries
import mlflow
import pandas as pd
import numpy as np
import matplotlib.pyplot as pyplot
from sklearn.model_selection import train_test_split, KFold
from sklearn.preprocessing import StandardScaler, PolynomialFeatures, LabelEncoder
from sklearn.metrics import mean_squared_error, r2_score
import joblib
import warnings
warnings.filterwarnings('ignore')
from sklearn.model_selection import KFold
import numpy as np
from sklearn.preprocessing import StandardScaler



In [2]:
# Set the tracking URI (local for now; adjust for server later)
mlflow.set_tracking_uri("file:///tmp/mlruns")

# Set or create an experiment
experiment_name = "car_price_prediction"  
mlflow.set_experiment(experiment_name)

print(f"MLflow tracking URI: {mlflow.get_tracking_uri()}")
print(f"Current experiment: {mlflow.get_experiment_by_name(experiment_name).name}")

MLflow tracking URI: file:///tmp/mlruns
Current experiment: car_price_prediction


In [3]:
class LinearRegression(object):

    """Linear Regression with Gradient Descent and Cross-Validation."""
    
    # We'll use the kfold object from the main script instead,
    # to avoid creating it inside the class.
    # We will keep this here just for default if the user does not provide it.
    kfold = KFold(n_splits=3, shuffle=False)

    def __init__(self,
                 regularization,
                 lr=0.001,
                 method='batch',          # 'batch' | 'mini' | 'sto'
                 num_epochs=1000,         # epochs for training
                 batch_size=50,           # batch size for mini-batch
                 cv=kfold,                # cross-validation object
                 init='zeros',            # weight initialization: 'zeros' | 'xavier'
                 use_momentum=False,      # momentum: T/F
                 momentum=0.9             # momentum values in (0,1)
                 ):
        self.lr             = lr
        self.num_epochs     = num_epochs
        self.batch_size     = batch_size
        self.method         = method
        self.cv             = cv
        self.regularization = regularization

        self.init          = init
        self.use_momentum  = use_momentum
        self.momentum      = momentum
        
        # public attributes to store training history for external use
        self.kfold_scores = []
        self.kfold_r2     = []
        self.theta = None # Initialize theta to None, it will be set in _init_theta

    def mse(self, ytrue, ypred):
        return ((ypred - ytrue) ** 2).sum() / ytrue.shape[0]

    def r2(self, ytrue, ypred):
        ss_res = np.sum((ytrue - ypred) ** 2)
        ss_tot = np.sum((ytrue - np.mean(ytrue)) ** 2)
        return 1.0 - ss_res / (ss_tot + 1e-12)

    def add_intercept(self, X):
        return np.c_[np.ones((X.shape[0], 1)), X]

    def _init_theta(self, n_features):
        """
        Initializes weights according to 'zeros' or 'xavier'.
        
        Params
        ------
        n_features: int
            Number of features (input dimensions).
        """
        # the number of inputs (m) is n_features - 1, since the first column is the bias.
        m = n_features - 1
        
        if self.init == 'xavier':
            # xavier uniform initialization: U[-1/sqrt(m), 1/sqrt(m)]
            limit = 1.0 / np.sqrt(m)
            self.theta = np.random.uniform(-limit, limit, size=(n_features,))
        else: # zeros initialization
            self.theta = np.zeros(n_features)

        # Initialize momentum buffer
        self._vtheta = np.zeros_like(self.theta)

    def fit(self, X_train, y_train):
        # Reset scores for each new fit call
        self.kfold_scores = []
        self.kfold_r2 = []

        X_train = np.asarray(X_train)
        y_train = np.asarray(y_train).reshape(-1)

        for fold, (train_idx, val_idx) in enumerate(self.cv.split(X_train)):
            X_cross_train = X_train[train_idx]
            y_cross_train = y_train[train_idx]
            X_cross_val   = X_train[val_idx]
            y_cross_val   = y_train[val_idx]
            
            # 1. Add intercept column to both training and validation sets
            X_cross_train = self.add_intercept(X_cross_train)
            X_cross_val   = self.add_intercept(X_cross_val)
            
            # 2. Scale only the feature columns (all but the first)
            scaler = StandardScaler()
            X_cross_train[:, 1:] = scaler.fit_transform(X_cross_train[:, 1:])
            X_cross_val[:, 1:]   = scaler.transform(X_cross_val[:, 1:])

            # Initialize weights
            self._init_theta(X_cross_train.shape[1])

            # CRITICAL CHANGE: The early stopping check has been removed.
            # The loop will now always run for the full number of epochs.
            for epoch in range(self.num_epochs):
                # Shuffle each epoch for SGD and mini-batch
                perm = np.random.permutation(X_cross_train.shape[0])
                Xtr_shuffled = X_cross_train[perm]
                ytr_shuffled = y_cross_train[perm]

                if self.method == 'sto':
                    for i in range(Xtr_shuffled.shape[0]):
                        X_method_train = Xtr_shuffled[i:i+1]
                        y_method_train = ytr_shuffled[i:i+1]
                        train_loss = self._train(X_method_train, y_method_train)
                elif self.method == 'mini':
                    for i in range(0, Xtr_shuffled.shape[0], self.batch_size):
                        X_method_train = Xtr_shuffled[i:i+self.batch_size]
                        y_method_train = ytr_shuffled[i:i+self.batch_size]
                        train_loss = self._train(X_method_train, y_method_train)
                else: # batch
                    train_loss = self._train(Xtr_shuffled, ytr_shuffled)

            yhat_val = self.predict(X_cross_val)
            final_val_loss = self.mse(y_cross_val, yhat_val)
            final_val_r2   = self.r2(y_cross_val, yhat_val)
            
            self.kfold_scores.append(final_val_loss)
            self.kfold_r2.append(final_val_r2)
            
            print(f"Fold {fold}: val_mse={final_val_loss:.6f}  val_r2={final_val_r2:.4f}")

    def _train(self, X, y):
        # Calculate prediction
        yhat = self.predict(X)
        m    = X.shape[0]

        # Calculate gradient for MSE
        grad = (1.0 / m) * X.T @ (yhat - y)
        
        # Add regularization penalty to feature coefficients only (theta[1:])
        reg_derivation = self.regularization.derivation(self.theta[1:])
        grad[1:] += reg_derivation

        if self.use_momentum:
            # Momentum update
            self._vtheta = self.momentum * self._vtheta + self.lr * grad
            self.theta = self.theta - self._vtheta
        else:
            # Vanilla gradient descent
            self.theta = self.theta - self.lr * grad

        return self.mse(y, yhat)

    def predict(self, X):
        """
        Predicts target values for a given input matrix X.
        It assumes X is a matrix of features, possibly without an intercept.
        """
        # Check if the input already has an intercept column.
        # This is the key fix to prevent double-adding the intercept.
        if X.shape[1] + 1 == len(self.theta):
            # If the number of columns is one less than theta's length,
            # it means we need to add an intercept.
            X_with_intercept = self.add_intercept(X)
        elif X.shape[1] == len(self.theta):
            # If the number of columns is the same as theta's length,
            # we assume it already has an intercept.
            X_with_intercept = X
        else:
            # Handle unexpected dimension mismatch gracefully
            raise ValueError(f"Input X has an unexpected number of columns. Expected {len(self.theta)-1} or {len(self.theta)}, but got {X.shape[1]}.")
            
        return X_with_intercept @ self.theta


Load dataset

In [4]:
# Load and preprocess the car dataset
df = pd.read_csv('data/Cars.csv')

Feature Engineering 

In [5]:
# clean non-numeric columns and drop irrelevant ones
# extract numeric values from string columns like '23.4 kmpl'
df['mileage'] = df['mileage'].str.extract('(\d+\.?\d*)').astype(float)
df['engine'] = df['engine'].str.extract('(\d+)').astype(float)
df['max_power'] = df['max_power'].str.extract('(\d+\.?\d*)').astype(float)
# Drop the 'torque' and 'name' columns for simplicity and to avoid complex preprocessing
df = df.drop(columns=['torque', 'name'])

# explicitly handle missing values for all columns
# fill missing numeric values with the mean
from sklearn.impute import SimpleImputer
numeric_columns = ['year', 'km_driven', 'mileage', 'engine', 'max_power', 'seats']
imputer_num = SimpleImputer(strategy='mean')
df[numeric_columns] = imputer_num.fit_transform(df[numeric_columns])

# for any remaining missing values (e.g., in categorical columns), use the most frequent value
categorical_columns = ['fuel', 'seller_type', 'transmission', 'owner']
imputer_cat = SimpleImputer(strategy='most_frequent')
df[categorical_columns] = imputer_cat.fit_transform(df[categorical_columns])

# a final check and removal of any remaining rows with NaN values.
df.dropna(inplace=True)

# encoding categorical columns using LabelEncoder
for col in categorical_columns:
    label_encoder = LabelEncoder()
    df[col] = label_encoder.fit_transform(df[col].astype(str))


Feature Selections and Scaling

In [6]:
# define features and target
feature_names = numeric_columns + categorical_columns
X = df[feature_names]
y = df['selling_price']

# split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# scale features using StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# create polynomial features for specific experiments
poly = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly = poly.fit_transform(X_train_scaled)
X_test_poly = poly.transform(X_test_scaled)

print(f"Training set: {X_train_scaled.shape}")
print(f"Test set: {X_test_scaled.shape}")
print(f"Polynomial features: {X_train_poly.shape}")

Training set: (6502, 10)
Test set: (1626, 10)
Polynomial features: (6502, 65)


In [7]:
class NoPenalty:
    """Represents no regularization penalty."""
    def __init__(self, l=0):
        self.l = l
    
    def derivation(self, theta):
        return np.zeros_like(theta)

class RidgePenalty:
    """Represents L2 (Ridge) regularization."""
    def __init__(self, l=0.1):
        self.l = l
    
    def derivation(self, theta):
        return self.l * 2 * theta

class LassoPenalty:
    """Represents L1 (Lasso) regularization."""
    def __init__(self, l=0.1):
        self.l = l
    
    def derivation(self, theta):
        # np.sign returns 0 for 0, which is correct for Lasso's subgradient
        return self.l * np.sign(theta)


In [8]:
# ==== Grid sweep specification ====
sweep_spec = {
    'penalty_kinds' : {
        'normal': NoPenalty,
        'ridge': RidgePenalty,
        'lasso': LassoPenalty
    },
    'momentum_flags': [True, False],
    'batch_modes'   : ['stochastic', 'mini', 'batch'],
    'weight_starts' : ['zeros', 'xavier'],
    'etas'          : [0.01, 0.001, 0.0001],
    'feature_modes' : ['normal', 'polynomial']
}

run_summaries = []
print("\nProcessing cross validation")

for pen_kind, PenaltyClass in sweep_spec['penalty_kinds'].items():
    for mom_on in sweep_spec['momentum_flags']:
        for batch_mode in sweep_spec['batch_modes']:
            for w_init in sweep_spec['weight_starts']:
                for eta in sweep_spec['etas']:
                    for feat_mode in sweep_spec['feature_modes']:
                        
                        # apply filters to skip invalid configurations
                        if (feat_mode == 'polynomial') and (pen_kind == 'normal'):
                            continue
                        if (feat_mode == 'polynomial') and (eta in {0.01, 0.001}):
                            continue

                        # Select the correct feature matrix
                        X_mat = X_train_poly if feat_mode == 'polynomial' else X_train_scaled
                        
                        run_name = f"{pen_kind}_{feat_mode}_mom{mom_on}_{batch_mode}_init{w_init}_lr{eta}"

                        with mlflow.start_run(run_name=run_name, nested=True):
                            mlflow.log_params({
                                "regularization": pen_kind,
                                "feature_type"  : feat_mode,
                                "use_momentum"  : mom_on,
                                "batch_type"    : batch_mode,
                                "init_method"   : w_init,
                                "learning_rate" : eta,
                                "lambda_reg"    : 0.1
                            })
                            
                            regularization_obj = PenaltyClass(l=0.1)
                            
                            estimator = LinearRegression(
                                regularization=regularization_obj,
                                lr=eta,
                                method=batch_mode,
                                use_momentum=mom_on,
                                init=w_init
                            )

                            try:
                                # The fit method handles the internal CV
                                estimator.fit(X_mat, y_train.values)
                                
                                # Use np.mean on the lists stored in the estimator
                                avg_cv_mse = np.mean(estimator.kfold_scores)
                                avg_cv_r2  = np.mean(estimator.kfold_r2)

                                mlflow.log_metric("avg_cv_r2",  float(avg_cv_r2))
                                mlflow.log_metric("avg_cv_mse", float(avg_cv_mse))

                                print(f"Run: {run_name} | Avg R2: {avg_cv_r2:.4f} | Avg MSE: {avg_cv_mse:.2f}")

                                run_summaries.append({
                                    'experiment_name': run_name,
                                    'mean_r2'        : avg_cv_r2,
                                    'mean_mse'       : avg_cv_mse,
                                    'config'         : {
                                        "regularization": pen_kind,
                                      
                                        "use_momentum"  : mom_on,
                                        "batch_type"    : batch_mode,
                                        "init_method"   : w_init,
                                        "learning_rate" : eta,
                                        "feature_type"  : feat_mode
                                    },
                                })

                            except Exception as ex:
                                print(f"Error in run {run_name}: {ex}")
                                mlflow.log_param("status", "Failed")
                                mlflow.log_param("error", str(ex))

                            mlflow.end_run()

print(f"\nFinished {len(run_summaries)} successful runs!")
df_results = pd.DataFrame(run_summaries)

if not df_results.empty:
    df_results = df_results.sort_values(by='mean_r2', ascending=False)
    
    # --- final Evaluation on Test Set ---
    # find the best configuration from the cross-validation results
    best_run = df_results.iloc[0]
    best_config = best_run['config']
    print(f"\nBest performing run based on CV: {best_run['experiment_name']}")
    
    # select the correct feature matrix for the best model
    if best_config['feature_type'] == 'polynomial':
        X_final_train = X_train_poly
        X_final_test = X_test_poly
    else:
        X_final_train = X_train_scaled
        X_final_test = X_test_scaled

    # instantiate the best model and train it on the full training set
    best_regularization = sweep_spec['penalty_kinds'][best_config['regularization']](l=0.1)
    final_model = LinearRegression(
        regularization=best_regularization,
        lr=best_config['learning_rate'],
        method=best_config['batch_type'],
        use_momentum=best_config['use_momentum'],
        init=best_config['init_method']
    )
    final_model.fit(X_final_train, y_train.values)
    
    # make final predictions on the test set
    final_predictions = final_model.predict(X_final_test)
    final_mse = mean_squared_error(y_test.values, final_predictions)
    final_r2 = final_model.r2(y_test.values, final_predictions)

    print(f"\n--- Final Test Set Results ---")
    print(f"Best Model's Hyperparameters: {best_config}")
    print(f"Final Test MSE: {final_mse:.4f}")
    print(f"Final Test R2: {final_r2:.4f}")
    
    # display top 10 results from the grid sweep
    print("\nTop 10 performing runs from cross-validation:")
    print(df_results.head(10))


Processing cross validation
Fold 0: val_mse=225887716519.962891  val_r2=0.6834
Fold 1: val_mse=198728696894.539581  val_r2=0.6542
Fold 2: val_mse=216431691124.385101  val_r2=0.6708
Run: normal_normal_momTrue_stochastic_initzeros_lr0.01 | Avg R2: 0.6695 | Avg MSE: 213682701512.96
Fold 0: val_mse=226324009676.516296  val_r2=0.6828
Fold 1: val_mse=198499921480.806854  val_r2=0.6546
Fold 2: val_mse=217215343488.764130  val_r2=0.6696
Run: normal_normal_momTrue_stochastic_initzeros_lr0.001 | Avg R2: 0.6690 | Avg MSE: 214013091548.70
Fold 0: val_mse=312605359127.938904  val_r2=0.5618
Fold 1: val_mse=251336986165.828369  val_r2=0.5626
Fold 2: val_mse=290969569464.044556  val_r2=0.5575
Run: normal_normal_momTrue_stochastic_initzeros_lr0.0001 | Avg R2: 0.5606 | Avg MSE: 284970638252.60
Fold 0: val_mse=225887716519.962799  val_r2=0.6834
Fold 1: val_mse=198728696894.538971  val_r2=0.6542
Fold 2: val_mse=216431691124.387085  val_r2=0.6708
Run: normal_normal_momTrue_stochastic_initxavier_lr0.01 | A

In [9]:
# --- Final Training on the Best Configuration ---
# best-performing hyperparameters from the cross-validation output
BEST_CONFIG = {
    "regularization": 'lasso',
    "use_momentum": True,
    "batch_type": 'mini',
    "init_method": 'xavier',
    "learning_rate": 0.0001,
    "feature_type": 'polynomial'
}

print(f"Training final model with best configuration: {BEST_CONFIG}")

# instantiate objects for scaling and polynomial features
scaler = StandardScaler()
poly = PolynomialFeatures(degree=2, include_bias=False)

# split data and apply correct transformations
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# normal features
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# polynomial features
X_train_poly = poly.fit_transform(X_train_scaled)
X_test_poly = poly.transform(X_test_scaled)

# select the correct feature matrix based on the best config
X_final_train = X_train_poly if BEST_CONFIG['feature_type'] == 'polynomial' else X_train_scaled

# instantiate the final model with the best parameters
best_regularization_obj = LassoPenalty(l=0.1)
final_model = LinearRegression(
    regularization=best_regularization_obj,
    lr=BEST_CONFIG['learning_rate'],
    method=BEST_CONFIG['batch_type'],
    use_momentum=BEST_CONFIG['use_momentum'],
    init=BEST_CONFIG['init_method'],
    cv=KFold(n_splits=2) 
)

# train the final model on the entire training set
final_model.fit(X_final_train, y_train.values)
final_predictions = final_model.predict(X_test_poly if BEST_CONFIG['feature_type'] == 'polynomial' else X_test_scaled)

Training final model with best configuration: {'regularization': 'lasso', 'use_momentum': True, 'batch_type': 'mini', 'init_method': 'xavier', 'learning_rate': 0.0001, 'feature_type': 'polynomial'}
Fold 0: val_mse=72030590599.930969  val_r2=0.8977
Fold 1: val_mse=65159964131.974731  val_r2=0.8901


In [10]:
import joblib
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, PolynomialFeatures

# For polynomial version
scaler = StandardScaler()
poly = PolynomialFeatures(degree=2, include_bias=False)

# Build a pipeline that applies scaling + poly + your custom LR
full_model = Pipeline([
    ("scaler", scaler),
    ("poly", poly),              # keep if BEST_CONFIG['feature_type'] == 'polynomial'
    ("est", final_model)
])

# Fit pipeline on raw features
full_model.fit(X_train, y_train)

# Save pipeline
joblib.dump(full_model, "app/artifacts/car-price-prediction-new.joblib")


Fold 0: val_mse=71169747124.168594  val_r2=0.8989
Fold 1: val_mse=65078766563.202141  val_r2=0.8902


['app/artifacts/car-price-prediction-new.joblib']

builtIn this report, I explain what I did when trying to tune the settings (hyperparameters) of a Linear Regression model to predict car prices. I modified the linear regression class of professor Chaky to include the r^2 and xavier weight initialization. To find the best setup, I ran a grid search where I tested different features, optimization methods, and parameter values. I used cross-validation to measure how well each setup worked. After that, I took the best model and tested it on a separate dataset that the model had never seen before, to check its real-world performance.

One of the most important things I learned was that adding polynomial features made a big difference. Without them, the models didn’t do well because they couldn’t capture the non-linear patterns in the data. Regularization also turned out to be really important, especially because polynomial features created a lot of new variables.

The best model I found had an average R² score of about 0.89 in cross-validation. Its best settings were:

Regularization: Lasso

Features: Polynomial

Optimization: Mini-batch gradient descent with Momentum

Initialization: Xavier

Learning Rate: 0.0001

When I tested this final model on the separate test set, I got:

Test MSE: 180,935,043,387.69

Test R²: 0.7240

The drop from 0.89 (cross-validation) to 0.72 (test set) was expected, and it showed me why having a test set is so important for fair evaluation. Overall, I was able to build a model that predicts car prices fairly accurately. The main factor for success was definitely using polynomial features.

<img src="best.png" alt="best output" width="500" height="300"/>

Best Output